builtin/logical/transit/path_decrypt.go

```diff
@@ -84,24 +109,114 @@ func (b *backend) pathDecryptWrite(
 		return logical.ErrorResponse("policy not found"), logical.ErrInvalidRequest
 	}
 
-	plaintext, err := p.Decrypt(context, nonce, ciphertext)
-	if err != nil {
-		switch err.(type) {
-		case errutil.UserError:
-			return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
-		case errutil.InternalError:
-			return nil, err
-		default:
-			return nil, err
+	batchInputRaw := d.Get("batch").(string)
+	var batchInputArray []BatchDecryptionItemRequest
+	var batchInput []byte
+	if len(batchInputRaw) != 0 {
+		batchInput, err = base64.StdEncoding.DecodeString(batchInputRaw)
+		if err != nil {
+			return logical.ErrorResponse("failed to base64-decode batch input"), logical.ErrInvalidRequest
+		}
+
+		if err := jsonutil.DecodeJSON([]byte(batchInput), &batchInputArray); err != nil {
+			return nil, fmt.Errorf("invalid input: %v", err)
+		}
+
+	} else {
+		ciphertext := d.Get("ciphertext").(string)
+		if len(ciphertext) == 0 {
+			return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+		}
+
+		batchInputArray = make([]BatchDecryptionItemRequest, 1)
+		batchInputArray[0] = BatchDecryptionItemRequest{
+			Ciphertext: ciphertext,
+			Context:    d.Get("context").(string),
+			Nonce:      d.Get("nonce").(string),
 		}
 	}
 
-	// Generate the response
-	resp := &logical.Response{
-		Data: map[string]interface{}{
-			"plaintext": plaintext,
-		},
+	var contextSet bool
+	switch len(batchInputArray) {
+	case 0:
+		return logical.ErrorResponse("missing input to process"), logical.ErrInvalidRequest
+	case 1:
+		contextSet = batchInputArray[0].Context != ""
+	default:
+		contextSet = batchInputArray[0].Context != ""
+		for _, item := range batchInputArray {
+			if (item.Context == "" && contextSet) || (item.Context != "" && !contextSet) {
+				return logical.ErrorResponse("context should be set either in all the request blocks or in none"), logical.ErrInvalidRequest
+			}
+		}
+	}
+
+	batchResponseItems := make([]BatchDecryptionItemResponse, len(batchInputArray))
+	for i, item := range batchInputArray {
+		if item.Ciphertext == "" {
```

Comment:
Can you move these three checks up into the `switch` above since you're already doing a `range` there, and then do the lock acquisition afterwards? That gets the verification logic out of the critical path.